service/src/main/java/bio/terra/axonserver/app/controller/GetFileController.java

```diff
@@ -64,16 +66,34 @@ public ResponseEntity<byte[]> getFile(
    * @return - A ResponseEntity containing the file
    */
   @Override
-  public ResponseEntity<byte[]> getFileInBucket(
+  public ResponseEntity<Resource> getFileInBucket(
       UUID workspaceId, UUID resourceId, String objectPath, @Nullable String convertTo) {
+    return buildResponse(workspaceId, resourceId, objectPath, convertTo);
+  }
 
+  private ResponseEntity<Resource> buildResponse(
+      UUID workspaceId, UUID resourceId, @Nullable String objectPath, @Nullable String convertTo) {
     BearerToken token = getToken();
 
     HttpRange byteRange = getByteRange();
 
-    byte[] resourceObj =
+    File resourceObj =
         fileService.getFile(token, workspaceId, resourceId, objectPath, convertTo, byteRange);
-    return new ResponseEntity<>(resourceObj, HttpStatus.OK);
+    String contentType =
+        convertTo == null
+            ? URLConnection.guessContentTypeFromName(resourceObj.getPath())
+            : URLConnection.guessContentTypeFromName((resourceObj.getName() + "." + convertTo));
+
+    HttpStatus resStatus = byteRange == null ? HttpStatus.OK : HttpStatus.PARTIAL_CONTENT;
+    HttpHeaders resHeaders = new HttpHeaders();
+    resHeaders.set(HttpHeaders.CONTENT_TYPE, contentType);
+
+    try {
+      InputStream inputStream = new FileInputStream(resourceObj);
+      return new ResponseEntity<>(new InputStreamResource(inputStream), resHeaders, resStatus);
+    } catch (IOException e) {
+      throw new InternalServerErrorException("Failed to read: " + resourceObj.getName());
```

Comment:
Wouldn't file not found be a possibility here? Seems like it should not be a 500 server error.

Reply:
By this point we already have the requested file, we just need create an input stream to pass it to the response. I'm wrapping the IOException that could happen but shouldn't happen unless there was an error with writing the fetched object locally, hence 500 error.